Core/MIPS/MIPSTracer.h

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2024- PPSSPP Project.
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 2.0 or later versions.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License 2.0 for more details.
+
+// A copy of the GPL 2.0 should have been included with the program.
+// If not, see http://www.gnu.org/licenses/
+
+// Official git repository and contact information can be found at
+// https://github.com/hrydgard/ppsspp and http://www.ppsspp.org/.
+
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+#include <string>
+
+#include "Common/CommonTypes.h"
+#include "Core/Opcode.h"
+#include "Core/MIPS/IR/IRJit.h"
+#include "Common/Log.h"
+#include "Common/File/Path.h"
+
+
+struct TraceBlockInfo {
+	u32 virt_address;
+	u32 storage_index;
+};
+
+struct TraceBlockStorage {
+	std::vector<u32> raw_instructions;
+	u32 cur_index;
+	u32* cur_data_ptr;
+
+	TraceBlockStorage(u32 capacity):
+		raw_instructions(capacity, 0),
+		cur_index(0),
+		cur_data_ptr(raw_instructions.data())
+		{}
+
+	TraceBlockStorage(): raw_instructions(), cur_index(0), cur_data_ptr(nullptr) {}
+
+	bool save_block(const u32* instructions, u32 size);
+
+	void initialize(u32 capacity);
+	void clear();
+
+	u32 operator[](u32 index) {
+		return raw_instructions[index];
+	}
+	Memory::Opcode read_asm(u32 index) {
+		return Memory::Opcode(raw_instructions[index]);
+	}
+};
+
+
+template <typename T>
+struct CyclicBuffer {
```

Comment:
As mentioned, i'd like to see this moved into a new header in Common/Data/Collections, though would also be fine with this happening in a separate PR later.